File: 0x00 基本算法/0x05 排序/AcWing105.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long LL;
const int N = 1e5 + 10;
int n, m, t, row[N], col[N];
LL s[N], c[N];

// 见 AcWing 122. 糖果传递
LL calc(int n, int a[]) {
  for (int i = 1; i <= n; i ++ ) s[i] = s[i - 1] + a[i];
  if (s[n] % n != 0) return -1; // 判断是否能够均分

  LL b = s[n] / n;
  int k = 0;
  for (int i = 1; i < n; i ++ ) c[k ++ ] = i * b - s[i];
  c[k ++ ] = 0;

  nth_element(c, c + k / 2, c + k); // sort(c, c + k);
  LL res = 0;
  for (int i = 0; i < n; i ++ ) res += abs(c[i] - c[k / 2]);
  return res;
}

int main () {
  cin >> n >> m >> t;
  while (t -- ) {
    int x, y;
    cin >> x >> y;
    row[x] ++ , col[y] ++ ;
  }

  LL res1 = calc(n, row), res2 = calc(m, col); // 行和列完全正交，互不干扰，分别计算
  if (res1 != -1 && res2 != -1) cout << "both " << res1 + res2 << endl;
  else if (res1 != -1) cout << "row " << res1 << endl;
  else if (res2 != -1) cout << "column " << res2 << endl;
  else puts("impossible");
  return 0;
}

File: 0x00 基本算法/0x05 排序/AcWing105.py
import sys

def calc(size, arr):
    total = sum(arr)
    if total % size != 0:
        return -1
        
    target = total // size
    c = [0] * size
    curr = 0
    for i in range(size - 1):
        curr += arr[i]
        c[i] = (i + 1) * target - curr
        
    c.sort()
    median = c[size // 2]
    return sum(abs(x - median) for x in c)

def main():
    input_data = sys.stdin.read().split()
    if not input_data:
        return
        
    iterator = iter(int(x) for x in input_data)
    
    n = next(iterator)
    m = next(iterator)
    t = next(iterator)
    
    row = [0] * n
    col = [0] * m
    
    for _ in range(t):
        row[next(iterator) - 1] += 1
        col[next(iterator) - 1] += 1
        
    res1 = calc(n, row)
    res2 = calc(m, col)
    
    if res1 != -1 and res2 != -1:
        print(f"both {res1 + res2}")
    elif res1 != -1:
        print(f"row {res1}")
    elif res2 != -1:
        print(f"column {res2}")
    else:
        print("impossible")

if __name__ == '__main__':
    main()